menuconfig.py

```diff
@@ -165,100 +165,274 @@
+    help=fg:cyan,bg:blue,bold
+    frame=fg:white,bg:cyan,bold
+    body=fg:white,bg:blue
+    edit=fg:black,bg:white
+    """
+}
 
-    # Initialize styles for different parts of the application. The arguments
-    # are ordered as follows:
-    #
-    #   1. Text color
-    #   2. Background color
-    #   3. Attributes
-    #   4. Extra attributes if colors aren't available. The colors will be
-    #      ignored in this case, and the attributes from (3.) and (4.) will be
-    #      ORed together.
+_CUSTOM_STYLE = None
 
-    # A_BOLD tends to produce faint and hard-to-read text on the Windows
-    # console, especially with the old color scheme, before the introduction of
-    # https://blogs.msdn.microsoft.com/commandline/2017/08/02/updating-the-windows-console-colors/
-    BOLD = curses.A_NORMAL if _IS_WINDOWS else curses.A_BOLD
+class StyleError(Exception):
+    pass
 
-    # Default styling. Themes can override these settings below.
+def _parse_style_single(style_str):
+    template_name = None
+    style = {}
 
-    # Top row in the main display, with the menu path
-    PATH_STYLE               = (curses.COLOR_BLACK, curses.COLOR_WHITE,  BOLD                              )
+    for sline in style_str.replace("\n", "").split(" "):
+        sline.strip(" ")
 
-    # Separator lines between windows. Also used for the top line in the symbol
-    # information dialog.
-    SEPARATOR_STYLE          = (curses.COLOR_BLACK, curses.COLOR_YELLOW, BOLD,            curses.A_STANDOUT)
+        # Skip empty entries
+        if len(sline) == 0: continue
 
-    # List of items, e.g. the main display
-    LIST_STYLE               = (curses.COLOR_BLACK, curses.COLOR_WHITE,  curses.A_NORMAL                   )
+        skey = sline.split("=")
+        # Words without a "=" character represents a style template
+        if len(skey) == 1:
+            template_name = skey[0]
 
-    # Style for the selected item
-    LIST_SEL_STYLE           = (curses.COLOR_WHITE, curses.COLOR_BLUE,   BOLD,            curses.A_STANDOUT)
+        # Key=data value, store it
+        elif len(skey) == 2:
+            style[skey[0]] = skey[1]
+        else:
+            raise StyleError("Style parse error near: {}".format(sline))
 
-    # Like _LIST_(SEL_)STYLE, for invisible items. Used in show-all mode.
-    LIST_INVISIBLE_STYLE     = (curses.COLOR_RED,   curses.COLOR_WHITE,  curses.A_NORMAL, BOLD             )
-    LIST_INVISIBLE_SEL_STYLE = (curses.COLOR_RED,   curses.COLOR_BLUE,   BOLD,            curses.A_STANDOUT)
+    return (style, template_name)
 
-    # Help text windows at the bottom of various fullscreen dialogs
-    HELP_STYLE               = PATH_STYLE
+def _parse_style(style_str=True):
+    tsyle = {}
+    tname = None
 
-    # Frame around dialog boxes
-    DIALOG_FRAME_STYLE       = SEPARATOR_STYLE
+    # Parse the current style, tname will hold the style template
+    # name
+    (style, tname) = _parse_style_single(style_str)
```

Comment:
There are still two functions, mainly because we have to deal with the "default" theme being the starting point of all themes.

I did change the algorithm according to your suggestion, with a slight modification. I believe that the following examples should produce the same style, but correct me if I'm wrong:

MENUCONFIG_STYLE="aquatic selection=fg:red,bg:black"
vs
MENUCONFIG_STYLE="selection=fg:red,bg:black aquatic"

The second statement should not overwrite the "selection" item from aquatic.

Anyways, it's possible to achieve quite funny results right now:

MENUCONFIG_STYLE="monochrome aquatic"

![2018-09-04-105235_1904x743_scrot](https://user-images.githubusercontent.com/1593069/45021014-c4b64b00-b030-11e8-99ef-301780aa981e.png)

Reply:
> There are still two functions, mainly because we have to deal with the "default" theme being the starting point of all themes.

Could handle that with a hardcoded `_parse_style("default")` before doing anything else perhaps.

> I did change the algorithm according to your suggestion, with a slight modification. I believe that the following examples should produce the same style, but correct me if I'm wrong:
> 
> MENUCONFIG_STYLE="aquatic selection=fg:red,bg:black"
> vs
> MENUCONFIG_STYLE="selection=fg:red,bg:black aquatic"

IMO, it's fine if they produce different results. Always having stuff further to the right override stuff on the left makes the behavior simple to understand and flexible.

I'd want to document that it works like that later. I could finish up the doc part though.

Reply:
I'll change it to what you suggested -- actually that makes the code even simpler.

Reply:
Fixed.